storm-core/src/clj/backtype/storm/daemon/nimbus.clj

```diff
@@ -346,53 +394,124 @@
       [(.getNodeId slot) (.getPort slot)]
       )))
 
+(defn- get-nimbus-subject []
+  (let [nimbus-subject (Subject.)
+        nimbus-principal (NimbusPrincipal.)
+        principals (.getPrincipals nimbus-subject)]
+    (.add principals nimbus-principal)
+    nimbus-subject))
+
+(defn- get-metadata-version [blob-store key subject]
+  (let [blob-meta (.getBlobMeta blob-store key subject)]
+    (.get_version blob-meta)))
+
+(defn get-key-list-from-blob-store [blob-store]
+  (let [key-iter (.listKeys blob-store (get-nimbus-subject))
+        keys (iterator-seq key-iter)]
+    (if (not-nil? keys)
+      (java.util.ArrayList. keys)
+      [])))
+
 (defn- setup-storm-code [nimbus conf storm-id tmp-jar-location storm-conf topology]
-  (let [stormroot (master-stormdist-root conf storm-id)]
-   (log-message "nimbus file location:" stormroot)
-   (FileUtils/forceMkdir (File. stormroot))
-   (FileUtils/cleanDirectory (File. stormroot))
-   (setup-jar conf tmp-jar-location stormroot)
-   (FileUtils/writeByteArrayToFile (File. (master-stormcode-path stormroot)) (Utils/serialize topology))
-   (FileUtils/writeByteArrayToFile (File. (master-stormconf-path stormroot)) (Utils/toCompressedJsonConf storm-conf))
-   (if (:code-distributor nimbus) (.upload (:code-distributor nimbus) stormroot storm-id))
-   ))
+  (let [subject (get-subject)
+        storm-cluster-state (:storm-cluster-state nimbus)
+        blob-store (:blob-store nimbus)
+        jar-key (master-stormjar-key storm-id)
+        code-key (master-stormcode-key storm-id)
+        conf-key (master-stormconf-key storm-id)
+        nimbus-host-port-info (:nimbus-host-port-info nimbus)]
+    (log-message "subject-changed" subject)
+    (if tmp-jar-location ;;in local mode there is no jar
+      (do
+        (log-message "tmp-jar-location" tmp-jar-location)
+        (.createBlob blob-store jar-key (FileInputStream. tmp-jar-location) (SettableBlobMeta. BlobStoreAclHandler/DEFAULT) subject)
+        (if (instance? LocalFsBlobStore blob-store)
+          (.setup-blobstore! storm-cluster-state jar-key nimbus-host-port-info (get-metadata-version blob-store jar-key subject)))
```

Comment:
Why is there a separate call to .setup-blobstore! here?  Shouldn't the blob store itself handle this?